File: app/src/main/java/com/patlejch/se1challenge/OrderGenerator.kt
package com.patlejch.se1challenge

import kotlin.random.Random
import java.util.UUID

interface OrderGenerator {
    fun next() : Order
}

class RandomOrderGenerator(_probability: Double): OrderGenerator {
    private var last: Order
    private val probability: Double
    
    init {
        last = createNewOrder()
        probability = _probability
    }
    
    override fun next() : Order {
        if (shouldCreateNewOrder()) {
            last = createNewOrder()
        }
        return last
    }

    private fun createNewOrder() = Order(UUID.randomUUID().toString())
    private fun shouldCreateNewOrder() = getRandomProbability() < probability
    private fun getRandomProbability() = Random.nextDouble(0.0, 1.0)
}

File: SE1ChallengeApp/Services/OrderGenerator.swift

import Foundation

protocol OrderGenerator {
    func next() -> Order
}

final class RandomOrderGenerator: OrderGenerator {
    private var last: Order!
    private let probability: Double
    
    init(probability: Double) {
        self.probability = probability
        last = createNewOrder()
    }
    
    func next() -> Order {
        if shouldCreateNewOrder() {
            last = createNewOrder()
        }
        return last
    }
    
    private func createNewOrder() -> Order { Order(id: UUID().uuidString) }
    private func shouldCreateNewOrder() -> Bool { getRandomProbability() < probability }
    private func getRandomProbability() -> Double { Double.random(in: 0.0 ... 1.0) }
}
